### PR TITLE
[mypyc] Make tuple packing and unpacking more efficient

### PR DESCRIPTION
**Pull request by JukkaL**
Previously returning a tuple from a function resulted in redundant increfs and decrefs
for each item, and similarly unpacking the returned tuple in an assignment had extra 
incref/decref pair per item. This PR introduces these changes to make this better:
* Creating a tuple steals the items always.
* Accessing a tuple item optionally borrows the item.
* A borrowed reference can be turned into a regular one using the new `Unborrow` op.
* The no-op `KeepAlive` op can steal the operands to avoid decrefing the operands.

Assignment from tuple now uses the three final features to avoid increfs and decrefs 
when unpacking a tuple in assignment. The docstrings in this PR contain additional
explanation of how this works.

In a micro-benchmark this improved performance by about 2-5%. In realistic examples 
the impact is likely small, but every little helps.

Here is an example where this helps:
```
def f() -> tuple[C, C]:
    return C(), C()  # Avoid 2 increfs and 2 decrefs

def g() -> None:
    x, y = f()  # Avoid 2 increfs and 2 decrefs
    ...
```

### CODE REVIEW

File: mypyc/ir/ops.py [hunk 1]
```diff
@@ -1387,21 +1391,76 @@ class KeepAlive(RegisterOp):
...
+        if self.steal:
+            return self.src.copy()
+        return []
+
     def accept(self, visitor: OpVisitor[T]) -> T:
         return visitor.visit_keep_alive(self)
 
 
+class Unborrow(RegisterOp):
+    """A no-op op to create a regular referece from a borrowed one.
```
Comment (AlexWaygood) [inline mypyc/ir/ops.py:1423]:
```suggestion
    """A no-op op to create a regular reference from a borrowed one.
```

File: mypyc/ir/ops.py [hunk 2]
```diff
@@ -1387,21 +1391,76 @@ class KeepAlive(RegisterOp):
...
+
+    Borrowed references can only be used temporarily and the reference
+    counts won't be managed. This value will be refcounted normally.
+
+    This is mainly useful if you split an aggregate value, such as
+    a tuple, into components using borrowed values (to avoid increfs),
+    and want to treat the components as sharing the original managed
+    reference. You'll also need to use KeepAlive with stel=True to
```
Comment (AlexWaygood) [inline mypyc/ir/ops.py:1431]:
```suggestion
    reference. You'll also need to use KeepAlive with steal=True to
```

File: mypyc/irbuild/statement.py [hunk 3]
```diff
@@ -183,8 +185,29 @@ def transform_assignment_stmt(builder: IRBuilder, stmt: AssignmentStmt) -> None:
 
     line = stmt.rvalue.line
     rvalue_reg = builder.accept(stmt.rvalue)
+
     if builder.non_function_scope() and stmt.is_final_def:
         builder.init_final_static(first_lvalue, rvalue_reg)
+
+    # Special case multiple assignments like 'x, y = expr' to reduce refcount ops.
```
Comment (AlexWaygood) [inline mypyc/irbuild/statement.py:192]:
```suggestion
    # Special-case multiple assignments like 'x, y = expr' to reduce refcount ops.
```

**Review by AlexWaygood — COMMENTED** [review]
One or two typos I spotted:

**Comment by JukkaL** [comment]
I'm not sure what's going on with the build failures. The tests seem to pass locally, but in CI the mypyc-compiled builds get a sigterm, apparently.

**Review by jhance — APPROVED** [review]

**Pull request merged**